### PR TITLE
logs/collapse_future_stack

### PR DESCRIPTION
**Pull request by chylli-binary**

### CODE REVIEW

**Review by shantanu-deriv — APPROVED** [review]

File: lib/Log/Any/Adapter/DERIV.pm [hunk 1]
```diff
@@ -251,6 +251,37 @@ sub log_entry {
...
+Return: log data
+
+=cut
+
+sub collapse_future_stack{
+    my ($self, $data) = @_;
+    my $stack = $data->{stack};
+    my @new_stack;
+    my $previous_future;
+    for my  $frame ($stack->@*){
```
Comment (shantanu-deriv) [inline lib/Log/Any/Adapter/DERIV.pm:268]: do this `for` loop guarantee that the latest `Future` stack trace will be available after collapsing?

File: t/02-stack.t [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+use Test::More;
+use Log::Any qw($log);
+use Log::Any::Adapter qw();
+use Log::Any::Adapter::DERIV;
+use Path::Tiny;
+use Future;
+use JSON::MaybeUTF8 qw(:v1);
+use Clone qw(clone);
+
+subtest 'collapse_future_stack' => sub {
```
Comment (shantanu-deriv) [inline t/02-stack.t:12]: I ran this subtest locally and found out that both `expected_stack` and `arg_stack` were empty so can we add an example where it is not empty?

File: lib/Log/Any/Adapter/DERIV.pm (hunk 1 above)
Comment (chylli-binary) [reply lib/Log/Any/Adapter/DERIV.pm:268]: It will keep the oldest one (outest), not the latest one (deepest). I'm not sure which one will be better. What do you think ?

File: t/02-stack.t (hunk 2 above)
Comment (chylli-binary) [reply t/02-stack.t:12]: Sorry ? I tried several times they shouldn't be empty. Let me try again.

File: lib/Log/Any/Adapter/DERIV.pm (hunk 1 above)
Comment (shantanu-deriv) [reply lib/Log/Any/Adapter/DERIV.pm:268]: Well, we should keep the one where the point of failure occurs. I can give an parallel example of null pointer exceptions (uninitialised value error).

```
my $var = undef;
function1($var);

#inside function1:
...
function2($var);
...

#inside function2:
...
function3($var);
...

#inside function3:
...
$var->{key} // point of failure as error occurs here
...
```

The stack trace could be something like
```
at line 29 in module3.pm in function3
at line 57 in module2.pm in function2
at line 17 in module1.pm in function1
at line8 in main.pl
```

I think both have their merits:
1. The topmost stack helps us understand what are the effects of the error (what will get impacted)
2. The deepest stack helps us understand the root cause

So whichever has more value can be used / preserved

File: t/02-stack.t (hunk 2 above)
Comment (shantanu-deriv) [reply t/02-stack.t:12]: I printed them both with `Date::Dumper` and both were empty (blank on STDOUT)

Comment (chylli-binary) [reply t/02-stack.t:12]: I don't know what command you used, but I'm sure your result is not correct. You can also check the logic. There are 3 items in the array of sample_stack,and both of them are the cloned data from it. Then `shift` will remove one item from `expected_stack`, then there are 2 items there. 

I guess you didn't add a `-l` when you run `prove` ?

File: lib/Log/Any/Adapter/DERIV.pm (hunk 1 above)
Comment (chylli-binary) [reply lib/Log/Any/Adapter/DERIV.pm:268]: So, in your conclusion, which one is better ?

Comment (sasikala-binary) [reply lib/Log/Any/Adapter/DERIV.pm:268]: I usually look for deepest first (the line which throws error) and then from where it  get called. stack trace normally last in first out

Comment (shantanu-deriv) [reply lib/Log/Any/Adapter/DERIV.pm:268]: yeah it is more helpful I feel

Comment (chylli-binary) [reply lib/Log/Any/Adapter/DERIV.pm:268]: Ok will fix it.

Comment (chylli-binary) [reply lib/Log/Any/Adapter/DERIV.pm:268]: Fix it @sasikala-binary @shantanu-deriv Please review again.
